<html><head><title>Sanger::CGP::Vaf::Process::Variant</title>
<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1" >
<link rel="stylesheet" title="blkbluw" type="text/css" href="../../../../_blkbluw.css" media="all" >
<link rel="alternate stylesheet" title="blkmagw" type="text/css" href="../../../../_blkmagw.css" media="all" >
<link rel="alternate stylesheet" title="blkcynw" type="text/css" href="../../../../_blkcynw.css" media="all" >
<link rel="alternate stylesheet" title="whtprpk" type="text/css" href="../../../../_whtprpk.css" media="all" >
<link rel="alternate stylesheet" title="whtnavk" type="text/css" href="../../../../_whtnavk.css" media="all" >
<link rel="alternate stylesheet" title="grygrnk" type="text/css" href="../../../../_grygrnk.css" media="all" >
<link rel="alternate stylesheet" title="whtgrng" type="text/css" href="../../../../_whtgrng.css" media="all" >
<link rel="alternate stylesheet" title="blkgrng" type="text/css" href="../../../../_blkgrng.css" media="all" >
<link rel="alternate stylesheet" title="grygrnw" type="text/css" href="../../../../_grygrnw.css" media="all" >
<link rel="alternate stylesheet" title="blkbluw" type="text/css" href="../../../../_blkbluw.css" media="all" >
<link rel="alternate stylesheet" title="whtpurk" type="text/css" href="../../../../_whtpurk.css" media="all" >
<link rel="alternate stylesheet" title="whtgrng" type="text/css" href="../../../../_whtgrng.css" media="all" >
<link rel="alternate stylesheet" title="grygrnw" type="text/css" href="../../../../_grygrnw.css" media="all" >

<script type="text/javascript" src="../../../../_podly.js"></script>

</head>
<body class='pod'>
<!--
  generated by Pod::Simple::HTML v3.20,
  using Pod::Simple::PullParser v3.20,
  under Perl v5.016003 at Tue May 23 11:09:35 2017 GMT.

 If you want to change this HTML document, you probably shouldn't do that
   by changing it directly.  Instead, see about changing the calling options
   to Pod::Simple::HTML, and/or subclassing Pod::Simple::HTML,
   then reconverting this document from the Pod source.
   When in doubt, email the author of Pod::Simple::HTML for advice.
   See 'perldoc Pod::Simple::HTML' for more info.

-->

<!-- start doc -->
<p class="backlinktop"><b><a name="___top" href="../../../../index.html" accesskey="1" title="All Documents">&lt;&lt;</a></b></p>

<div class='indexgroup'>
<ul   class='indexList indexList1'>
  <ul   class='indexList indexList2'>
    <li class='indexItem indexItem2'><a href='#formatVarinat_get_hash_containing_location_specific_information_Inputs_%3Dover_2_%3Dback'>formatVarinat get hash containing location specific information Inputs =over 2 =back</a>
    <li class='indexItem indexItem2'><a href='#_fetch_reads_fetch_reads'>_fetch_reads fetch reads</a>
    <li class='indexItem indexItem2'><a href='#addNormalCount_add_count_for_normal_sample_Inputs_%3Dover_2_%3Ditem_g_pu_-_has_containing_sample_specific_info_%3Dback'>addNormalCount add count for normal sample Inputs =over 2 =item g_pu - has containing sample specific info =back</a>
  </ul>
</ul>
</div>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="formatVarinat_get_hash_containing_location_specific_information_Inputs_=over_2_=back"
>formatVarinat get hash containing location specific information Inputs =over 2 =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>formatVarinat get hash containing location specific information Inputs =over 2 =item store_results -store results for printing =item g_pu -hash containing pileup/exonerate based allele fractions =item sample -sample name =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>getVcfFields get sample specific original values for INFO and FILTER fields Inputs =over 2 =item data_for_all_samples -original VCF fields for a given sample =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_getFlagStatus get new filter flag value based on original filter defined in VCF Inputs =over 2 =item flag_val -Original flag values in filter column =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_get_INFO parse info field Inputs =over 2 =item original_vcf_info -data from INFO field =item DNFS -Descriptive flag status [ref header info] =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>createExonerateInput Create input files for exonerate Inputs =over 2 =item bam -bam object =item bam_header -bam header data =item depth -max depth at this location =item g_pu -hash containing sample specific info =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_getRange get left and right span from indel position Inputs =over 2 =item bam_header -sample specific information on read length,
chr length and lib size =item g_pu -hash containing sample specific info =item max_depth -max depth at this location =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_check_hdr_overlap checks overlap with high depth regions Inputs =over 2 =item chr -chromosome number =item zero_start start coordiante =item one_end -end coordinate =tabix - tabix object of bed file containing UCSC high depth regions =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_get_dna_segment get reference dna Inputs =over 2 =item bam_object - Bio::DB sam object =item chr - chromosome number =item start - start position =item end - end position =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_get_alt_seq get alternative reference seq Inputs =over 2 =item bam_objects - Bio::DB sam object =item g_pu - hash containing sample specific info =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_get_ref_5p_pos get updated 5p positions Inputs =over 2 =item ref_seq -reference sequence =item reconstructed_alt_seq &#226;&#128;&#148;alternate sequence =item g_pu -stores relative positions of variant region =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>populateHash populate has for chr start and end Inputs =over 2 =item g_pu - has containing sample specific info =item sample - sample name =item bam_header_data - sample specific information on read length and lib size =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_fetch_features get reads in an given location using Bio:Db:Sam fetch method Inputs =over 2 =item sam_object - Bio::DB sam object =item g_pu - hash containing sample specific info =item Reads_FH - file handler to store reads for a given location =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_fetch_features get reads in an given location using Bio:Db:Sam fetch method Inputs =over 2 =item sam_object - Bio::DB sam object =item g_pu - hash containing sample specific info =item Reads_FH - file handler to store reads for a given location =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="_fetch_reads_fetch_reads"
>_fetch_reads fetch reads</a></h2>

<p>Inputs =over 2 =item sam_object - Bio::DB sam object =item region - chr:start-stop format region info to get reads =item Reads_FH - temp file handler to store reads =back</p>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_fetch_mate_seq get mate sequence Inputs =over 2 =item sam_object - Bio::DB sam object =item region - postion to get reads =item readname - mate readname =item Reads_FH - file handler to store read sequence =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_fetch_unmapped_reads fetch reads whose mate is unmapped Inputs =over 2 =item sam_object - Bio::DB sam object =item region - chr:start-stop format region info to get reads =item Reads_FH - temp file handler to store reads =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_do_exonerate parse exonerate output Inputs =over 2 =item ref_seq_file &#226;&#128;&#148;reference sequence and alt sequences in fasta format =item temp_read_file &#226;&#128;&#148;temp read file in fasta format =item g_pu -stores relative positions of variant region =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_cleaup_read_ambiguities cleaup of ambiguous reads and read of a same read pair mapping on +ve and/or -ve strand at same location Inputs =over 2 =item g_pu -- stores relative positions of variant region =item read_track_alt &#226;&#128;&#148;hash stroring reads mapping on alt base =item read_track_ref &#226;&#128;&#148;hash stroring reads mapping on ref base =item alt_count_p &#226;&#128;&#148;hash stroring +ve reads mapping on alt base =item alt_count_n &#226;&#128;&#148;hash stroring -ve reads mapping on alt base =item ref_count_p &#226;&#128;&#148;hash stroring +ve reads mapping on ref base =item ref_count_n &#226;&#128;&#148;hash stroring -ve reads mapping on ref base =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="addNormalCount_add_count_for_normal_sample_Inputs_=over_2_=item_g_pu_-_has_containing_sample_specific_info_=back"
>addNormalCount add count for normal sample Inputs =over 2 =item g_pu - has containing sample specific info =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_get_pileup get pileup output for given location Inputs =over 2 =item bam_object - Bio::DB sam object =item g_pu - hash containing sample specific info =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>formatResults format pileup/ exonerate results as per VCF specifications =over 2 =item original_flag -orignal flag values =item g_pu -hash containing results and sample specific info for give location =back</a></h2>
<p class="backlinkbottom"><b><a name="___bottom" href="../../../../index.html" title="All Documents">&lt;&lt;</a></b></p>

<!-- end doc -->

</body></html>
