<html><head><title>Sanger::CGP::Vaf::Data::ReadVcf</title>
<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1" >
<link rel="stylesheet" title="blkbluw" type="text/css" href="../../../../_blkbluw.css" media="all" >
<link rel="alternate stylesheet" title="blkmagw" type="text/css" href="../../../../_blkmagw.css" media="all" >
<link rel="alternate stylesheet" title="blkcynw" type="text/css" href="../../../../_blkcynw.css" media="all" >
<link rel="alternate stylesheet" title="whtprpk" type="text/css" href="../../../../_whtprpk.css" media="all" >
<link rel="alternate stylesheet" title="whtnavk" type="text/css" href="../../../../_whtnavk.css" media="all" >
<link rel="alternate stylesheet" title="grygrnk" type="text/css" href="../../../../_grygrnk.css" media="all" >
<link rel="alternate stylesheet" title="whtgrng" type="text/css" href="../../../../_whtgrng.css" media="all" >
<link rel="alternate stylesheet" title="blkgrng" type="text/css" href="../../../../_blkgrng.css" media="all" >
<link rel="alternate stylesheet" title="grygrnw" type="text/css" href="../../../../_grygrnw.css" media="all" >
<link rel="alternate stylesheet" title="blkbluw" type="text/css" href="../../../../_blkbluw.css" media="all" >
<link rel="alternate stylesheet" title="whtpurk" type="text/css" href="../../../../_whtpurk.css" media="all" >
<link rel="alternate stylesheet" title="whtgrng" type="text/css" href="../../../../_whtgrng.css" media="all" >
<link rel="alternate stylesheet" title="grygrnw" type="text/css" href="../../../../_grygrnw.css" media="all" >

<script type="text/javascript" src="../../../../_podly.js"></script>

</head>
<body class='pod'>
<!--
  generated by Pod::Simple::HTML v3.20,
  using Pod::Simple::PullParser v3.20,
  under Perl v5.016003 at Tue May 23 11:09:35 2017 GMT.

 If you want to change this HTML document, you probably shouldn't do that
   by changing it directly.  Instead, see about changing the calling options
   to Pod::Simple::HTML, and/or subclassing Pod::Simple::HTML,
   then reconverting this document from the Pod source.
   When in doubt, email the author of Pod::Simple::HTML for advice.
   See 'perldoc Pod::Simple::HTML' for more info.

-->

<!-- start doc -->
<p class="backlinktop"><b><a name="___top" href="../../../../index.html" accesskey="1" title="All Documents">&lt;&lt;</a></b></p>

<div class='indexgroup'>
<ul   class='indexList indexList1'>
  <ul   class='indexList indexList2'>
    <li class='indexItem indexItem2'><a href='#getVcfHeaderData_get_header_info_from_vcf_file_Inputs_%3Dover_2_%3Dback'>getVcfHeaderData get header info from vcf file Inputs =over 2 =back</a>
    <li class='indexItem indexItem2'><a href='#_checkNormal_check_normal_sample_Inputs_%3Dover_2_%3Ditem_vcf_normal_-normal_sample_defined_in_vcf_header_%3Dback'>_checkNormal check normal sample Inputs =over 2 =item vcf_normal -normal sample defined in vcf header =back</a>
    <li class='indexItem indexItem2'><a href='#getProgress_get_analysis_progress_log_Inputs_%3Dover_2_%3Dback'>getProgress get analysis progress log Inputs =over 2 =back</a>
    <li class='indexItem indexItem2'><a href='#_populateBedHeader_add_bed_info_to_vcf_header_Inputs_%3Dover_2_%3Ditem_info_tag_val_-vcf_info_tag_object_%3Dback'>_populateBedHeader add bed info to vcf header Inputs =over 2 =item info_tag_val -vcf info tag object =back</a>
    <li class='indexItem indexItem2'><a href='#getChromosomes_get_chromosome_names_from_genome_file_Inputs_%3Dover_2_%3Dback'>getChromosomes get chromosome names from genome file Inputs =over 2 =back</a>
    <li class='indexItem indexItem2'><a href='#_getData_check_and_get_user_provided_data_and_populate_respective_hash_values_Inputs_%3Dover_2_%3Dback'>_getData check and get user provided data and populate respective hash values Inputs =over 2 =back</a>
    <li class='indexItem indexItem2'><a href='#_getProcessLog_get_process_log_to_write_in_vcf_header_Inputs_%3Dover_2_%3Dback'>_getProcessLog get process log to write in vcf header Inputs =over 2 =back</a>
    <li class='indexItem indexItem2'><a href='#_trim_file_path_trim_path_to_write_in_process_log_Inputs_%3Dover_2_%3Ditem_string_-full_path_string'>_trim_file_path trim path to write in process log Inputs =over 2 =item string -full path string</a>
    <li class='indexItem indexItem2'><a href='#getBedHash_create_has_for_user_defined_bed_locations_Inputs_%3Dover_2_%3Dback'>getBedHash create has for user defined bed locations Inputs =over 2 =back</a>
    <li class='indexItem indexItem2'><a href='#_get_bam_object_create_bam_object_using_Bio%3A%3ADB%3A%3ASam_Inputs_%3Dover_2_%3Dback'>_get_bam_object create bam object using Bio::DB::Sam Inputs =over 2 =back</a>
    <li class='indexItem indexItem2'><a href='#_get_lib_size_from_bas_get_library_size_from_BAS_file_Inputs_%3Dover_2_%3Ditem_bas_file_-bas_file_name_%3Dback'>_get_lib_size_from_bas get library size from BAS file Inputs =over 2 =item bas_file -bas file name =back</a>
    <li class='indexItem indexItem2'><a href='#_get_read_length_get_read_length_per_sample_Inputs_%3Dover_2_%3Ditem_sam_-_Bio%3A%3ADB%3A%3ASam_objects_%3Dback'>_get_read_length get read length per sample Inputs =over 2 =item sam - Bio::DB::Sam objects =back</a>
    <li class='indexItem indexItem2'><a href='#WriteAugmentedHeader_write_VCF_header_data_Inputs_%3Dover_2_%3Dback'>WriteAugmentedHeader write VCF header data Inputs =over 2 =back</a>
    <li class='indexItem indexItem2'><a href='#getVCFObject_write_VCF_header_data_Inputs_%3Dover_2_%3Ditem_info_tag_val_-VCF_INFO_tag_values_%3Dback'>getVCFObject write VCF header data Inputs =over 2 =item info_tag_val -VCF INFO tag values =back</a>
    <li class='indexItem indexItem2'><a href='#_get_tab_sep_header_parse_header_data_to_generate_header_data_and_columns_names_Inputs_%3Dover_2_%3Ditem_vcf_header_%3Dback'>_get_tab_sep_header parse header data to generate header data and columns names Inputs =over 2 =item vcf_header =back</a>
    <li class='indexItem indexItem2'><a href='#gzipAndIndexVcf_Write_bgzip_compressed_vcf_file_Inputs_%3Dover_2_%3Ditem_annot_vcf_file_to_compress_%3Dback'>gzipAndIndexVcf Write bgzip compressed vcf file Inputs =over 2 =item annot_vcf file to compress =back</a>
    <li class='indexItem indexItem2'><a href='#_runCommand_run_external_command_Inputs_%3Dover_2_%3Ditem_command_to_run_%3Dback'>_runCommand run external command Inputs =over 2 =item command to run =back</a>
    <li class='indexItem indexItem2'><a href='#catFiles_cat_files_Inputs_%3Dover_2_%3Ditem_path_-file_path_%3Ditem_ext_-file_extension_%3Ditem_outfile_-outfile_name_%3Dback'>catFiles cat files Inputs =over 2 =item path -file path =item ext -file extension =item outfile -outfile name =back</a>
    <ul   class='indexList indexList3'>
      <li class='indexItem indexItem3'><a href='#clear_path'>clear_path</a>
    </ul>
  </ul>
  <li class='indexItem indexItem1'><a href='#POD_ERRORS'>POD ERRORS</a>
</ul>
</div>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="getVcfHeaderData_get_header_info_from_vcf_file_Inputs_=over_2_=back"
>getVcfHeaderData get header info from vcf file Inputs =over 2 =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="_checkNormal_check_normal_sample_Inputs_=over_2_=item_vcf_normal_-normal_sample_defined_in_vcf_header_=back"
>_checkNormal check normal sample Inputs =over 2 =item vcf_normal -normal sample defined in vcf header =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="getProgress_get_analysis_progress_log_Inputs_=over_2_=back"
>getProgress get analysis progress log Inputs =over 2 =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="_populateBedHeader_add_bed_info_to_vcf_header_Inputs_=over_2_=item_info_tag_val_-vcf_info_tag_object_=back"
>_populateBedHeader add bed info to vcf header Inputs =over 2 =item info_tag_val -vcf info tag object =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>writeFinalFileHeaders create sample specific file handlers to write VAF output in tsv and vcf format Inputs =over 2 =item info_tag_val -vcf info tag object =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="getChromosomes_get_chromosome_names_from_genome_file_Inputs_=over_2_=back"
>getChromosomes get chromosome names from genome file Inputs =over 2 =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>getMergedLocations get merged snp/indel locations for a given individual group Inputs =over 2 =item =item chr_location -chromosome or an interval on chromosome =item updated_info_tags -vcf info tag object =item vcf_file_obj -vcf file object =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="_getData_check_and_get_user_provided_data_and_populate_respective_hash_values_Inputs_=over_2_=back"
>_getData check and get user provided data and populate respective hash values Inputs =over 2 =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_getOriginalHeader parse VCF header and get data from old header lines Inputs =over 2 =item vcf - vcf object =item info_tag_val - hash to store data as perl VCF header specifications declared in get_unique_locations so that samples names for each VCF were stored =item info_tags -user defined annotation tags to be included in the INFO field [e.g Vagrent annotations] =item norma_sample -name of the normal sample to be stored =item tumour_count -counter to increment tumour sample name in header =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_getCustomHeader stores custom VCF header information as per VCF specifications Inputs =over 2 =item vcf_filter - hash created to store FILTER field =item vcf_info - hash created to store INFO field =item vcf_foramt - hash created to store FORMAT field =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="_getProcessLog_get_process_log_to_write_in_vcf_header_Inputs_=over_2_=back"
>_getProcessLog get process log to write in vcf header Inputs =over 2 =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="_trim_file_path_trim_path_to_write_in_process_log_Inputs_=over_2_=item_string_-full_path_string"
>_trim_file_path trim path to write in process log Inputs =over 2 =item string -full path string</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_getInfo parse VCF INFO line for given locations and returns values for respective tags Inputs =over 2 =item vcf - vcf object =item INFO - VCF INFO filed value =item INFO - INFO field tags defined in header =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_getReadDepth get read depth for given variant site for underlying sample line[8] is format field,
get index of tag used for depth e.g.,
NR(reads on -ve strand) and PR (reads on +ve strand) once we had index get value for given tag in the tumour sample Inputs =over 2 =item vcf -- vcf object =item line -- vcf line for give location =item tags --user specified tags that account for total read depth =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="getBedHash_create_has_for_user_defined_bed_locations_Inputs_=over_2_=back"
>getBedHash create has for user defined bed locations Inputs =over 2 =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>filterBedLocations Filter bed locations present in vcf file Inputs =over 2 =item unique_locations -union of locations created by merging vcf files =item bed_locations -hash of bed locations =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>populateBedLocations Populate bed locations for all the samples Inputs =over 2 =item filtered_bed_locations -Filtered bed locations present in vcf file =item updated_info_tags -INFO tags for bed locations =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>processMergedLocations Analyse merged vcf and/or bed locations Inputs =over 2 =item data_for_all_samples -sample specific location information =item unique_locations -merged unique locations for each individual =item variant -varinat varinat object containing location specific variant information =item bam_header_data -header info from BAM file =item bam_objects -Bio::DB sam object =item store_results -store results for given chromosome =item chr -chromosome =item tags -custom tags =item info_tag_value -hash containing info tag value =item progress_fhw -file handler to write progress data =item progress_data -progress data =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="_get_bam_object_create_bam_object_using_Bio::DB::Sam_Inputs_=over_2_=back"
>_get_bam_object create bam object using Bio::DB::Sam Inputs =over 2 =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_get_bam_header_data get_bam_header_data -insert size and chr length Inputs =over 2 =item bam_objects -Bio::DB sam object =item bas_files -bas file to get lib size =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="_get_lib_size_from_bas_get_library_size_from_BAS_file_Inputs_=over_2_=item_bas_file_-bas_file_name_=back"
>_get_lib_size_from_bas get library size from BAS file Inputs =over 2 =item bas_file -bas file name =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="_get_read_length_get_read_length_per_sample_Inputs_=over_2_=item_sam_-_Bio::DB::Sam_objects_=back"
>_get_read_length get read length per sample Inputs =over 2 =item sam - Bio::DB::Sam objects =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="WriteAugmentedHeader_write_VCF_header_data_Inputs_=over_2_=back"
>WriteAugmentedHeader write VCF header data Inputs =over 2 =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="getVCFObject_write_VCF_header_data_Inputs_=over_2_=item_info_tag_val_-VCF_INFO_tag_values_=back"
>getVCFObject write VCF header data Inputs =over 2 =item info_tag_val -VCF INFO tag values =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="_get_tab_sep_header_parse_header_data_to_generate_header_data_and_columns_names_Inputs_=over_2_=item_vcf_header_=back"
>_get_tab_sep_header parse header data to generate header data and columns names Inputs =over 2 =item vcf_header =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_get_header_lines parse header data to generate tab file header and columns names Inputs =over 2 =item header_line -- specific header line to parse =item val -- parse specific value from header =item prefix -header prefix e.g.,
SAMPLE,
FILTER,
FORMAT etc.,
=back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_writeOutput Write output to files Inputs =over 2 =item original_vcf_info - original VCF data =item NFS - New Filter status =item new_pileup_results --pileup/exonerate results for given position =item tags -custom tags =item WFH_VCF -Vcf Write file handler =item WFH_TSV -Tsv Write file handler =item g_pu -hash containing pielup/exonerate data =item vcf -VCF object =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_parse_info_line parse info line to generate tab sep outfile Inputs =over 2 =item vcf_line vcf --line to be parsed =item original_vcf_info --VCF info field =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>writeResults Write augmented vcf file Inputs =over 2 =item aug_vcf_fh -Augmented vcf file handler =item store_results -Hash containing results =item aug_vcf_name -Augmented vcf name =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'

>_writeFinalVcf Write augmented vcf file Inputs =over 2 =item vcf_file -vcf file name =item aug_vcf_fh -Hash of augmented vcf file handlers =item sample -sample name =item store_results -Hash containing results =item aug_vcf_name -Augmented vcf file name hash =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="gzipAndIndexVcf_Write_bgzip_compressed_vcf_file_Inputs_=over_2_=item_annot_vcf_file_to_compress_=back"
>gzipAndIndexVcf Write bgzip compressed vcf file Inputs =over 2 =item annot_vcf file to compress =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="_runCommand_run_external_command_Inputs_=over_2_=item_command_to_run_=back"
>_runCommand run external command Inputs =over 2 =item command to run =back</a></h2>

<h2><a class='u' href='#___top' title='click to go to top of document'
name="catFiles_cat_files_Inputs_=over_2_=item_path_-file_path_=item_ext_-file_extension_=item_outfile_-outfile_name_=back"
>catFiles cat files Inputs =over 2 =item path -file path =item ext -file extension =item outfile -outfile name =back</a></h2>

<h3><a class='u' href='#___top' title='click to go to top of document'
name="clear_path"
>clear_path</a></h3>

<blockquote>
<p>Implemented as File::Path remove_tree (rmtree) the idea is to remove the content of all of the directories and then use remove_tree to remove the directories needed to handle very big structures from devil</p>
</blockquote>

<h1><a class='u' href='#___top' title='click to go to top of document'
name="POD_ERRORS"
>POD ERRORS</a></h1>

<p>Hey!
<b>The above document had some coding errors,
which are explained below:</b></p>

<dl>
<dt><a name="Around_line_550:"
>Around line 550:</a></dt>

<dd>
<p>=back without =over</p>
</dd>
</dl>
<p class="backlinkbottom"><b><a name="___bottom" href="../../../../index.html" title="All Documents">&lt;&lt;</a></b></p>

<!-- end doc -->

</body></html>
