<html><head><title>cgpVaf</title>
<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1" >
<link rel="stylesheet" title="blkbluw" type="text/css" href="_blkbluw.css" media="all" >
<link rel="alternate stylesheet" title="blkmagw" type="text/css" href="_blkmagw.css" media="all" >
<link rel="alternate stylesheet" title="blkcynw" type="text/css" href="_blkcynw.css" media="all" >
<link rel="alternate stylesheet" title="whtprpk" type="text/css" href="_whtprpk.css" media="all" >
<link rel="alternate stylesheet" title="whtnavk" type="text/css" href="_whtnavk.css" media="all" >
<link rel="alternate stylesheet" title="grygrnk" type="text/css" href="_grygrnk.css" media="all" >
<link rel="alternate stylesheet" title="whtgrng" type="text/css" href="_whtgrng.css" media="all" >
<link rel="alternate stylesheet" title="blkgrng" type="text/css" href="_blkgrng.css" media="all" >
<link rel="alternate stylesheet" title="grygrnw" type="text/css" href="_grygrnw.css" media="all" >
<link rel="alternate stylesheet" title="blkbluw" type="text/css" href="_blkbluw.css" media="all" >
<link rel="alternate stylesheet" title="whtpurk" type="text/css" href="_whtpurk.css" media="all" >
<link rel="alternate stylesheet" title="whtgrng" type="text/css" href="_whtgrng.css" media="all" >
<link rel="alternate stylesheet" title="grygrnw" type="text/css" href="_grygrnw.css" media="all" >

<script type="text/javascript" src="_podly.js"></script>

</head>
<body class='pod'>
<!--
  generated by Pod::Simple::HTML v3.20,
  using Pod::Simple::PullParser v3.20,
  under Perl v5.016003 at Tue May 23 11:09:35 2017 GMT.

 If you want to change this HTML document, you probably shouldn't do that
   by changing it directly.  Instead, see about changing the calling options
   to Pod::Simple::HTML, and/or subclassing Pod::Simple::HTML,
   then reconverting this document from the Pod source.
   When in doubt, email the author of Pod::Simple::HTML for advice.
   See 'perldoc Pod::Simple::HTML' for more info.

-->

<!-- start doc -->
<p class="backlinktop"><b><a name="___top" href="index.html" accesskey="1" title="All Documents">&lt;&lt;</a></b></p>

<div class='indexgroup'>
<ul   class='indexList indexList1'>
  <li class='indexItem indexItem1'><a href='#NAME'>NAME</a>
  <li class='indexItem indexItem1'><a href='#SYNOPSIS'>SYNOPSIS</a>
</ul>
</div>

<h1><a class='u' href='#___top' title='click to go to top of document'
name="NAME"
>NAME</a></h1>

<p>cgpVaf.pl merge the variants in vcf files for a given Tumour - normal pairs in a project and write pileup and exonerate output for merged locations</p>

<h1><a class='u' href='#___top' title='click to go to top of document'
name="SYNOPSIS"
>SYNOPSIS</a></h1>

<p>cgpVaf.pl [-h] -d -a -g -tn -nn -e -o [ -b -t -c -r -m -ao -mq -pid -bo -vcf -v]</p>

<pre>  Required Options (inputDir and variant_type must be defined):

   --variant_type   (-a)  variant type (snp or indel) [default snp]
   --inputDir       (-d)  input directory path
   --genome         (-g)  genome fasta file name (default genome.fa)
   --tumour_name    (-tn) Toumour sample name [ list of space separated  sample names ]
   --normal_name    (-nn) Normal sample name [ single sample used as normal for this analysis ]
   --outDir         (-o)  Output folder
   --vcfExtension   (-e)  vcf file extension string after the sample name - INCLUDE&#39;s preceding dot (default: .caveman_c.annot.vcf.gz) 
 
  Optional 
   --infoTags       (-t)  comma separated list of tags to be included in the vcf INFO field 
                         (default: VD,VW,VT,VC for Vagrent annotations)
   --bedIntervals   (-b)  tab separated file containing list of intervals in the form of &#60;chr&#62;&#60;pos&#62; &#60;ref&#62;&#60;alt&#62; (e.g 1  14000  A  C)
                         bed file can be specified in the config file after the last sample pair,
                         if specified on command line then same bed file is used for all the tumour/normal pairs,
                         bed file name in config file overrides command line argument
   --hdr_cutoff     (-c)  High Depth Region(HDR) cutoff  value[ avoids extreme depth regions (default: 005 i.e top 0.05% )]
                         (possible values 001,005,01,05 and 1)
   --restrict_flag  (-r)  restrict analysis on (possible values 1 : PASS or 0 : ALL) [default 1 ]   
   --augment        (-m)  Augment pindel file [ this will add additional fields[ MTR, WTR, AMB] to FORMAT column of NORMAL and TUMOUR samples ] (default 0: don not augment)
   --augment_only   (-ao) Only augment pindel VCF file (-m must be specified) [ do not  merge input files and add non passed varinats to output file ] (default 0: augment and merge )
   --map_quality    (-mq) read mapping quality threshold
   --base_quality   (-bq) base quality threshold for snp
   --bamExtension   (-be) Input read file extension  
   --depth          (-dp) comma separated list of field(s) as specified in FORMAT field representing total depth at given location
   --high_depth_bed (-hdr) High Depth Region(HDR) bed file to mask high depth regions in the genome
   --id_int_project (-pid)Internal project id [WTSI only]
   --bed_only       (-bo) Only analyse bed intervals in the file (default 0: analyse vcf and bed interval)
   --vcf            (-vcf)user defined vcf file name (otherwise deduced from tumour name and  [please specify in same order as tumour sample names]
   --help           (-h)  Display this help message
   --version        (-v)  provide version information for vaf

   Examples:
      Merge vcf files to create single vcf containing union of all the variant sites and provides pileup output for each location
      perl cgpVaf.pl -d tmpvcfdir -o testout -a snp -g genome.fa -e .caveman_c.annot.vcf.gz -nn PD21369b -tn PD26296a PD26296c2 
      Merge vcf files to create single vcf containing union of all the variant sites and provides allele count for underlying indel location
      perl cgpVaf.pl -d tmpvcfdir -o testout -a indel -g genome.fa -e .caveman_c.annot.vcf.gz -nn PD21369b -tn PD26296a PD26296c2</pre>
<p class="backlinkbottom"><b><a name="___bottom" href="index.html" title="All Documents">&lt;&lt;</a></b></p>

<!-- end doc -->

</body></html>
